zigpy/application.py

```diff
@@ -191,71 +205,104 @@ def handle_message(
-            and cluster != 0
         ):
-            # only allow access to basic cluster while initializing endpoints
-            LOGGER.debug(
-                "Received frame on uninitialized device %s endpoint %s for cluster: %s",
-                sender.ieee,
-                dst_ep,
+            # Allow the following responses:
+            #  - any ZDO
+            #  - ZCL if endpoints are initialized
+            #  - ZCL from Basic cluster if endpoints are initializing
+
+            if not sender.initializing:
+                sender.schedule_initialize()
+
+            return sender.handle_message(
+                profile,
                 cluster,
+                src_ep,
+                dst_ep,
+                message,
+                dst_addressing=dst_addressing,
             )
-            return
-        return sender.handle_message(
-            profile,
-            cluster,
-            src_ep,
-            dst_ep,
-            message,
-            dst_addressing=dst_addressing,
+
+        # Give quirks a chance to fast-initialize the device (at the moment only Xiaomi)
+        zigpy.quirks.handle_message_from_uninitialized_sender(
+            sender, profile, cluster, src_ep, dst_ep, message
         )
 
-    def handle_join(self, nwk, ieee, parent_nwk):
+        # Reload the sender device object, in it was replaced by the quirk
+        sender = self.get_device(ieee=sender.ieee)
+
+        # If the quirk did not fast-initialize the device, start initialization
+        if not sender.initializing and not sender.is_initialized:
+            sender.schedule_initialize()
+
+    def handle_join(self, nwk: t.NWK, ieee: t.EUI64, parent_nwk: t.NWK) -> None:
+        """
+        Called when a device joins or announces itself on the network.
+        """
+
         ieee = t.EUI64(ieee)
-        LOGGER.info("Device 0x%04x (%s) joined the network", nwk, ieee)
-        if ieee in self.devices:
+
+        try:
             dev = self.get_device(ieee)
-            if dev.nwk != nwk:
-                LOGGER.debug(
-                    "Device %s changed id (0x%04x => 0x%04x)", ieee, dev.nwk, nwk
-                )
-                dev.nwk = nwk
-                dev.schedule_group_membership_scan()
-            elif dev.initializing or dev.status == zigpy.device.Status.ENDPOINTS_INIT:
-                LOGGER.debug("Skip initialization for existing device %s", ieee)
-                dev.schedule_group_membership_scan()
-                return
-        else:
+            LOGGER.info("Device 0x%04x (%s) joined the network", nwk, ieee)
+        except KeyError:
             dev = self.add_device(ieee, nwk)
+            LOGGER.info("New device 0x%04x (%s) joined the network", nwk, ieee)
+
+        if dev.nwk != nwk:
+            dev.nwk = nwk
+            LOGGER.debug("Device %s changed id (0x%04x => 0x%04x)", ieee, dev.nwk, nwk)
 
         self.listener_event("device_joined", dev)
+
         dev.schedule_initialize()
```

Comment:
HA `dev` works without any modifications 🎉  but the only regression right now is this line causing ZHA to do its re-initialization for every device that re-joins the network after a power cycle.

Zigpy didn't previously emit a `device_joined` or `device_initialized` event when an already-initialized device re-joined the network either. Maybe we need to change it to this?

```Python
if not dev.is_initialized:
    dev.schedule_initialize()
```

Also, does ZHA keep no initialization state of its own? As in, once a device joins from zigpy's perspective, is the ZHA initialization progress persisted anywhere?

Reply:
ZHA does not keep the initialization state of its own and cannot really tell a re-join from join, so it does ZHA device reconfigure every time it receives the event "device_joined"

Reply:
Also, not all radios can tell zigpy if it is a new join or a re-join. zigpy could have device in the DB, but it still is a new join if the device was reset and paired, without removing from zigpy 1st. I think we can assume, any time the device changes NWK it is a new join

Reply:
IMO we should do line 290 `self.listener_event("device_joined", dev)` only if NWK was changed or device did not exist before. Need to check ZHA code, as I believe it kicks of ZHA device configuration after the event fired by `Device.initialize()`

Reply:
For partially-initialized devices that finish initialization when receiving a message, how about doing something like `schedule_initialize(suppress_event=True)`? Or would it be better for ZHA to redo its initialization??

Reply:
Does it do a re-initialization for every device even the ones there were fully initialized before? I.e. have node descriptor, endpoints and clusters?

Reply:
It doesn't do anything but `schedule_initialize`/`initialize` will still call `self._application.device_initialized(self)` in the end.

Reply:
If it happens only for partially initialized devices, then I'd say that correct behavior. Would be mitigated on zha side by ignoring partially initialized devices

Reply:
I think if we `schedule_initialize()` only on:
1. Partially initialized device
2. Nwk change
3. New device created/no existing devices

Which in turn would trigger device re-initialization on zha side, then this would be correct and expected behavior.

If it's a handle_join() but NWK is the same and device is initialized, then we should skip schedule _initialize and just do group_member_scan()

Reply:
`handle_join` should now match this behavior.